### PR TITLE
JAVA-1196: Include hash of result set metadata in prepared statement id.

### PR DESCRIPTION
**Pull request by tolbertam**
Created from [ifesdjeen/java-driver#10786-release](https://github.com/ifesdjeen/java-driver/tree/10786-release) and rebased on 3.x.    Used [tolbertam/cassandra#10786-queryflags](https://github.com/tolbertam/cassandra/tree/10786-queryflags) to test it.

Still needs review, but the tests are running good against that C* branch.

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 1]
```diff
@@ -2201,7 +2201,14 @@ public PreparedStatement addPrepared(PreparedStatement stmt) {
 
                 // The one object in the cache will get GCed once it's not referenced by the client anymore since we use a weak reference.
                 // So we need to make sure that the instance we do return to the user is the one that is in the cache.
-                return previous;
+                if (previous.getPreparedId().getResultMetadataId().equals(stmt.getPreparedId().metadata)) {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: This should be ` if (previous.getPreparedId().getResultMetadataId().equals(stmt.getPreparedId().getResultMetadataId()))`.

File: driver-core/src/main/java/com/datastax/driver/core/PreparedId.java [hunk 2]
```diff
@@ -24,16 +24,33 @@
...
+    PreparedId(MD5Digest id, MD5Digest resultMetadataId, ColumnDefinitions metadata, ColumnDefinitions resultSetMetadata, int[] routingKeyIndexes, ProtocolVersion protocolVersion) {
         this.id = id;
+        this.resultMetadataId = resultMetadataId;
         this.metadata = metadata;
         this.resultSetMetadata = resultSetMetadata;
         this.routingKeyIndexes = routingKeyIndexes;
         this.protocolVersion = protocolVersion;
     }
+
+    public void swap(MD5Digest metadataId, ColumnDefinitions resultSetMetadata)
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/PreparedId.java]: This swap is not atomic, we should either synchronize on `this` or create an `AtomicReference` to hold a value object containing both data.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java (hunk 1 above)
Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: The warning line 2199 should imo only be printed if we are in this side of the if/else block.

File: driver-core/src/main/java/com/datastax/driver/core/Responses.java [hunk 3]
```diff
@@ -380,6 +383,14 @@ static Metadata decode(ByteBuf body, boolean withPkIndices, ProtocolVersion prot
                     EnumSet<Flag> flags = Flag.deserialize(body.readInt());
                     int columnCount = body.readInt();
 
+                    MD5Digest resultMetadataId = null;
+                    if (flags.contains(Flag.METADATA_CHANGED))
+                    {
+                        assert protocolVersion == ProtocolVersion.V5 : "MetadataChanged flag is supported starting from v5";
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/Responses.java]: Nit: `MetadataChanged` -> `METADATA_CHANGED`

**Review by adutra — COMMENTED** [review]
A couple of minor concerns + an overall problem of code formatting.

**Comment by adutra** [comment]
Went ahead and implemented my suggestions, and also refactored `PreparedId` a bit. Ready to review.

**Comment by adutra** [comment]
Squashed and rebased.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 4]
```diff
@@ -2208,16 +2208,35 @@ public void ensurePoolsSizing() {
         }
 
         public PreparedStatement addPrepared(PreparedStatement stmt) {
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java:2220]: Do we still need to change this method?
It's not very important to handle race conditions here; even if we store an instance that has the wrong `resultSetMetadata`, the server will inform us in the first response and we'll mutate the instance then.

**Review by olim7t — CHANGES_REQUESTED** [review]
Looks great overall, I think the cache update can be simplified.
One thing I wonder is if a paging query can be affected if the metadata changes between two pages, I'm going to test that.

**Comment by olim7t** [comment]
It is indeed possible to observe a metadata change between two pages of the same `ArrayBackedResultSet.MultiPage`.
I think we should handle that, it will require a bit of refactoring to make the metadata mutable, and ensure that we make the switch at the right time.

**Comment by olim7t** [comment]
> It is indeed possible to observe a metadata change between two pages of the same ArrayBackedResultSet.MultiPage

I'm working on a fix for this.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java (hunk 4 above)
Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java:2220]: Indeed, but at least the new version would not print a log warning about re-preparing the same statement, if it's being re-prepared because the resultset metadata has changed.

**Comment by adutra** [comment]
👍 lgtm

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java (hunk 4 above)
Comment (olim7t) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java:2220]: > if it's being re-prepared because the resultset metadata has changed.

This is no longer the case. When the resultset metadata has changed, we don't re-prepare and re-add the statement, only mutate it in place.
So the only way `addPrepared` is called twice is if the client explicitly called `session.prepare` twice with the same statement. Whether the resultset metadata changed or not between the two calls is irrelevant I think, this is still an anti-pattern.

**Review by olim7t — APPROVED** [review]

**Review by adutra — APPROVED** [review]

File: driver-core/src/main/java/com/datastax/driver/core/SessionManager.java [hunk 5]
```diff
@@ -567,16 +566,23 @@ else if (fetchSize != Integer.MAX_VALUE)
...
             if (protocolVersion.compareTo(ProtocolVersion.V4) < 0)
                 bs.ensureAllSet();
-            boolean skipMetadata = protocolVersion != ProtocolVersion.V1 && bs.statement.getPreparedId().resultSetMetadata != null;
-            Requests.QueryProtocolOptions options = new Requests.QueryProtocolOptions(Message.Request.Type.EXECUTE, consistency, Arrays.asList(bs.wrapper.values), Collections.<String, ByteBuffer>emptyMap(),
-                    skipMetadata, fetchSize, usedPagingState, serialConsistency, defaultTimestamp);
-            request = new Requests.Execute(bs.statement.getPreparedId().id, options, statement.isTracing());
+
+            // skip resultset metadata if version > 1 (otherwise this feature is not supported)
+            // and if we already have metadata for the prepared statement being executed.
+            boolean skipMetadata = protocolVersion != ProtocolVersion.V1 && bs.statement.getPreparedId().resultSetMetadata.variables != null;
```
Comment (newkek) [inline driver-core/src/main/java/com/datastax/driver/core/SessionManager.java:586]: Isn't the driver supposed to always receive results metadata after a `PREPARE` message? I was wondering if the check for `bs.statement.getPreparedId().resultSetMetadata.variables != null` wasn't redundant.

File: driver-core/src/main/java/com/datastax/driver/core/Responses.java [hunk 6]
```diff
@@ -380,6 +384,17 @@ static Metadata decode(ByteBuf body, boolean withPkIndices, ProtocolVersion prot
                     EnumSet<Flag> flags = Flag.deserialize(body.readInt());
                     int columnCount = body.readInt();
 
+                    ByteBuffer state = null;
+                    if (flags.contains(Flag.HAS_MORE_PAGES))
+                        state = CBUtil.readValue(body);
+
+                    MD5Digest resultMetadataId = null;
+                    if (flags.contains(Flag.METADATA_CHANGED)) {
+                        assert protocolVersion == ProtocolVersion.V5 : "METADATA_CHANGED flag is supported starting from v5";
```
Comment (newkek) [inline driver-core/src/main/java/com/datastax/driver/core/Responses.java]: Is it possible that we would ever receive from Cassandra the `METADATA_CHANGED` with protocol < V5? Maybe that check isn't necessary?

Comment (ifesdjeen) [reply driver-core/src/main/java/com/datastax/driver/core/Responses.java]: This is the point of the assertion: we can not receive it. If we somehow did - throw assertion error.

**Comment by ifesdjeen** [comment]
It looks like the `cluster` is holding the old metadata, even after the session reconnect and table destruction, so the tests weren't idempotent (but which also means that technically re-creating the table with the same exact name might cause trouble). I've added `cluster` instance reset for now, but I think we should take a closer look at why that might be happening.

**Comment by ifesdjeen** [comment]
Added a fix for the problem described above. This is mostly relevant/realistic for tests, but theoretically may happen in real life as well.

When the following sequence of events happens:
  * create a cluster; create a table with schema `schema`;
  * prepare statement (receive an id of `hash1` / `resultset1`)
  * add a column 
  * try running a `SELECT` query
  * ensure that the metadata part of the id is now `resultset2`

So far, this is pretty much the issue description (all good). Now, the interesting part starts: 

  * drop the table; re-create the table (with the same `schema` as the initial one); 
  * try preparing the statement again, you'll notice that the metadata part of the id is again `resultset2`

Of course the resultset metadata should have been `resultset1`. 

The problem is that (I assume) for performance reasons, [here](https://github.com/datastax/java-driver/pull/794/commits/a4620a4c3eb66d0e3e4ad06e80564b57360744d7#diff-3c61240d382b855ddcdc43fa82420d28R2227) the "old" statement is returned (comment mentions since it's in cache, so I assume the CPU cache is meant. Problem is that while it was correct in the previous setup: same statement id == same statements, it's not true anymore, since the metadata might have changed. I've added a check to compare metadata of the returning statement to make sure we return the correct one and skip the optimisation when it is not possible to use it.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 7]
```diff
@@ -1388,7 +1388,7 @@ private static String generateClusterName() {
         // new one join the cluster).
         // Note: we could move this down to the session level, but since prepared statement are global to a node,
         // this would yield a slightly less clear behavior.
-        ConcurrentMap<MD5Digest, PreparedStatement> preparedQueries;
+        public ConcurrentMap<MD5Digest, PreparedStatement> preparedQueries;
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: Looks unrelated to the latest change, is it a leftover?

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 8]
```diff
@@ -2218,14 +2218,15 @@ public void ensurePoolsSizing() {
...
+            PreparedStatement previous = preparedQueries.putIfAbsent(stmt.getPreparedId().boundValuesMetadata.id, stmt);
             if (previous != null) {
                 logger.warn("Re-preparing already prepared query is generally an anti-pattern and will likely affect performance. "
                         + "Consider preparing the statement only once. Query='{}'", stmt.getQueryString());
 
                 // The one object in the cache will get GCed once it's not referenced by the client anymore since we use a weak reference.
                 // So we need to make sure that the instance we do return to the user is the one that is in the cache.
-                return previous;
+                if (com.google.common.base.Objects.equal(previous.getPreparedId().resultSetMetadata.id, stmt.getPreparedId().resultSetMetadata.id))
+                    return previous;
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: I don't think it's the right way to fix the issue you described. We're returning the correct PreparedStatement to the caller of this method, but leaving the cached instance with the old result metadata. Rather it should be:
```java
previous.getPreparedId().resultSetMetadata = stmt.getPreparedId().resultSetMetadata;
return previous;
```

**Review by olim7t — CHANGES_REQUESTED** [review]

File: driver-core/src/main/java/com/datastax/driver/core/ArrayBackedResultSet.java [hunk 9]
```diff
@@ -62,15 +62,34 @@ static ArrayBackedResultSet fromMessage(Responses.Result msg, SessionManager ses
...
+                    // only ever sets the flag SKIP_METADATA to true for bound statements,
+                    // never for regular ones.
                     Statement actualStatement = statement;
                     if (statement instanceof StatementWrapper) {
                         actualStatement = ((StatementWrapper) statement).getWrappedStatement();
                     }
-                    assert actualStatement instanceof BoundStatement;
-                    columnDefs = ((BoundStatement) actualStatement).statement.getPreparedId().resultSetMetadata;
+                    assert statement instanceof BoundStatement;
+                    columnDefs = ((BoundStatement) statement).statement.getPreparedId().resultSetMetadata.variables;
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/ArrayBackedResultSet.java]: Should be `actualStatement` on those two lines.

File: driver-core/src/main/java/com/datastax/driver/core/ArrayBackedResultSet.java [hunk 10]
```diff
@@ -62,15 +62,34 @@ static ArrayBackedResultSet fromMessage(Responses.Result msg, SessionManager ses
...
                     assert columnDefs != null;
                 } else {
+                    // metadata is always present for regular statements (QUERY requests);
+                    // it might also be present for bound statements (EXECUTE requests)
+                    // if the metadata has changed server-side.
                     columnDefs = r.metadata.columns;
+                    if (statement instanceof BoundStatement) {
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/ArrayBackedResultSet.java]: We should unwrap like in the other branch.

**Review by olim7t — CHANGES_REQUESTED** [review]
Sorry, I submitted my previous review too soon :(
Two more comments.

**Review by olim7t — APPROVED** [review]

**Comment by olim7t** [comment]
I went ahead and implemented my suggestions.

**Comment by ifesdjeen** [comment]
Thank you a lot @olim7t!

**Comment by olim7t** [comment]
Added `ProtocolFeature` enum and rebased on current 3.x.

File: driver-core/src/main/java/com/datastax/driver/core/ProtocolFeature.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Determines whether or not the input version supports ths feature.
+     *
+     * @param version the version to test against.
+     * @return true if supported, false otherwise.
+     */
+    boolean isSupportedBy(ProtocolVersion version) {
+        switch (this) {
+            case PREPARED_METADATA_CHANGES:
+                return version == ProtocolVersion.V5;
```
Comment (ifesdjeen) [inline driver-core/src/main/java/com/datastax/driver/core/ProtocolFeature.java:40]: Might be worth checking for `>`, will be hard to catch when  we bump the version again...

**Comment by ifesdjeen** [comment]
Are we waiting for a follow-up Cassandra ticket to be resolved?.. It'd be good to get it committed at some point, as we might have more v5 changes.

**Comment by olim7t** [comment]
Rebased on 3.3.x + changed base + fixed issues with conditional updates.
